File: apps/desktop/src/api/types/records.ts
import type { AssetRefs, PoolKind, SortDirection } from "./base";

export type DisplayRecord = {
  record_id: string;
  source_order: number;
  record_type: string;
  time?: string | null;
  pool_kind: PoolKind;
  pool_id: string;
  pool_label: string;
  banner: ResolvedBanner;
  item_id: string;
  item_name: string;
  item_asset_refs: AssetRefs;
  item_kind: ItemKind;
  rarity?: number | null;
  count?: number | null;
  roll_points?: number | null;
  roll_label_id?: string | null;
  roll_label?: string | null;
  roll_bucket: RollBucket;
  fork_result_mark?: ForkResultMark | null;
  secondary_item_id?: string | null;
  secondary_item_name?: string | null;
  secondary_item_asset_refs: AssetRefs;
  secondary_count?: number | null;
  derived: RecordDerived;
};

type BannerResolutionIssue =
  | "unknown_pool"
  | "unknown_time"
  | "outside_known_windows"
  | "ambiguous";

type RuleResolutionIssue =
  | "fallback_pool_kind"
  | "missing_banner"
  | "missing_rule"
  | "unsupported_scope";

export type RateUpResult = "up" | "off_rate" | "not_applicable" | "unknown";
export type PityBadge = "fork_up_guarantee" | "fork_5star_guarantee" | "fork_4star_guarantee";
export type ForkResultMark = "win" | "guaranteed" | "lose";
export type RollBucket = "gift" | "sleep" | "1" | "2" | "3" | "4" | "5" | "6" | "not_applicable";
export type ItemKind = "character" | "fork" | "appearance" | "inventory" | "vehicle_module" | "unknown";

export type ResolvedBanner = {
  resolution_issue?: BannerResolutionIssue | null;
  reason?: string | null;
  banner_id?: string | null;
  pool_id?: string | null;
  pool_kind?: PoolKind | string | null;
  banner_type?: "limited" | "standard" | "fork" | string | null;
  title?: string | null;
  version?: string | null;
  start_at?: string | null;
  end_at?: string | null;
  timezone?: string | null;
  rate_up_5: string[];
  rate_up_4: string[];
  standard_5_pool: string[];
  standard_4_pool: string[];
  rule_id?: string | null;
  asset_refs: AssetRefs;
};

export type GachaRuleView = {
  resolution_issue?: RuleResolutionIssue | null;
  reason?: string | null;
  rule_id?: string | null;
  pool_kind: PoolKind;
  hard_pity_5?: number | null;
  hard_up_pity_5?: number | null;
  pickup_win_rate_5?: number | null;
  has_guarantee_5?: boolean | null;
  guarantee_scope?: string | null;
  carry_scope?: string | null;
};

export type RecordDerived = {
  record_id: string;
  banner_id?: string | null;
  banner_version?: string | null;
  counts_as_pull: boolean;
  global_pull_no?: number | null;
  pull_no_in_pool_kind?: number | null;
  pull_no_in_banner?: number | null;
  pity_5_before: number;
  pity_5_after: number;
  ten_pull_progress_before?: number | null;
  ten_pull_progress_after?: number | null;
  hit_rarity?: number | null;
  rate_up_result: RateUpResult;
  pity_badge?: PityBadge | null;
  guarantee_5_before?: boolean | null;
  guarantee_5_after?: boolean | null;
  fork_up_pity_before?: number | null;
  fork_up_pity_after?: number | null;
  fork_forced_up?: boolean | null;
  rule: GachaRuleView;
};

export type RecordFilter = {
  pool_kind?: PoolKind | null;
  banner_ids?: string[] | null;
  rarities?: number[] | null;
  focused_rarities?: number[] | null;
  rate_up_results?: RateUpResult[] | null;
  roll_buckets?: RollBucket[] | null;
  item_kinds?: ItemKind[] | null;
  fork_result_marks?: ForkResultMark[] | null;
  fork_pity_badges?: PityBadge[] | null;
  date_from?: string | null;
  date_to?: string | null;
  search?: string | null;
  sort_direction?: SortDirection | null;
  limit?: number;
  offset?: number;
};

export type RecordList = {
  total: number;
  records: DisplayRecord[];
};

type RecordBannerOption = {
  banner_id: string;
  pool_kind: PoolKind;
  title: string;
  count: number;
};

type RecordRollBucketOption = {
  bucket: RollBucket;
  count: number;
};

type RecordItemKindOption = {
  item_kind: ItemKind;
  count: number;
};

export type RecordFilterOptions = {
  banners: RecordBannerOption[];
  roll_buckets: RecordRollBucketOption[];
  item_kinds: RecordItemKindOption[];
};

File: crates/nte-core/src/model/records.rs
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum PityBadge {
    #[serde(rename = "fork_up_guarantee")]
    ForkUpGuarantee,
    #[serde(rename = "fork_5star_guarantee", alias = "fork_five_star_guarantee")]
    ForkFiveStarGuarantee,
    #[serde(rename = "fork_4star_guarantee", alias = "fork_four_star_guarantee")]
    ForkFourStarGuarantee,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ForkResultMark {
    Win,
    Guaranteed,
    Lose,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RecordDerived {
    pub record_id: String,
    pub banner_id: Option<String>,
    pub banner_version: Option<String>,
    pub counts_as_pull: bool,
    pub global_pull_no: Option<u64>,
    pub pull_no_in_pool_kind: Option<u64>,
    pub pull_no_in_banner: Option<u64>,
    pub pity_5_before: u64,
    pub pity_5_after: u64,
    pub ten_pull_progress_before: Option<u8>,
    pub ten_pull_progress_after: Option<u8>,
    pub hit_rarity: Option<u8>,
    pub rate_up_result: RateUpResult,
    pub pity_badge: Option<PityBadge>,
    pub guarantee_5_before: Option<bool>,
    pub guarantee_5_after: Option<bool>,
    pub fork_up_pity_before: Option<u64>,
    pub fork_up_pity_after: Option<u64>,
    pub fork_forced_up: Option<bool>,
    pub rule: GachaRuleView,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct DisplayRecord {
    pub record_id: String,
    pub source_order: u64,
    pub record_type: String,
    pub time: Option<String>,
    pub pool_kind: PoolKind,
    pub pool_id: String,
    pub pool_label: String,
    pub banner: ResolvedBanner,
    pub item_id: String,
    pub item_name: String,
    pub item_asset_refs: BTreeMap<String, serde_json::Value>,
    pub item_kind: ItemKind,
    pub rarity: Option<u8>,
    pub count: Option<i64>,
    pub roll_points: Option<i64>,
    pub roll_label_id: Option<String>,
    pub roll_label: Option<String>,
    pub roll_bucket: RollBucket,
    pub fork_result_mark: Option<ForkResultMark>,
    pub secondary_item_id: Option<String>,
    pub secondary_item_name: Option<String>,
    pub secondary_item_asset_refs: BTreeMap<String, serde_json::Value>,
    pub secondary_count: Option<i64>,
    pub derived: RecordDerived,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
pub struct RecordFilter {
    pub pool_kind: Option<PoolKind>,
    #[serde(default)]
    pub banner_ids: Vec<String>,
    #[serde(default)]
    pub rarities: Vec<u8>,
    #[serde(default)]
    pub focused_rarities: Vec<u8>,
    #[serde(default)]
    pub rate_up_results: Vec<RateUpResult>,
    #[serde(default)]
    pub roll_buckets: Vec<RollBucket>,
    #[serde(default)]
    pub item_kinds: Vec<ItemKind>,
    #[serde(default)]
    pub fork_result_marks: Vec<ForkResultMark>,
    #[serde(default)]
    pub fork_pity_badges: Vec<PityBadge>,
    pub date_from: Option<String>,
    pub date_to: Option<String>,
    pub search: Option<String>,
    pub sort_direction: Option<SortDirection>,
    pub limit: Option<u32>,
    pub offset: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RecordList {
    pub total: u64,
    pub records: Vec<DisplayRecord>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RecordFilterOptions {
    pub banners: Vec<RecordBannerOption>,
    pub roll_buckets: Vec<RecordRollBucketOption>,
    pub item_kinds: Vec<RecordItemKindOption>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RecordBannerOption {
    pub banner_id: String,
    pub pool_kind: PoolKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolution_issue: Option<BannerResolutionIssue>,
    pub title: String,
    pub count: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RecordRollBucketOption {
    pub bucket: RollBucket,
    pub count: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RecordItemKindOption {
    pub item_kind: ItemKind,
    pub label: String,
    pub count: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct MapLocaleList {
    pub locales: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct AssetsPackManifest {
    pub schema: String,
    pub schema_version: u32,
    pub app_version: String,
    pub map_hash: String,
    pub source_repo: String,
    pub source_commit: String,
    pub format: String,
    pub quality: u8,
    pub file_count: u64,
    pub assets: Vec<AssetsPackAsset>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct AssetsPackAsset {
    pub asset_ref: String,
    pub kind: String,
    pub source_path: String,
    pub pack_path: String,
    pub width: u32,
    pub height: u32,
    pub sha256: String,
}
